src/Components/Server/src/Circuits/RemoteRenderer.cs

```diff
@@ -116,79 +134,179 @@ protected override Task UpdateDisplayAsync(in RenderBatch batch)
-
-            var pendingRenderInfo = new AutoCancelTaskCompletionSource<object>(TimeoutMilliseconds);
-            _pendingRenders[renderId] = pendingRenderInfo;
-
             // Send the render batch to the client
-            // If the "send" operation fails (synchronously or asynchronously), abort
-            // the whole render with that exception
-            try
+            // If the "send" operation fails (synchronously or asynchronously) or the client
+            // gets disconected retry three times and then give up. This likely mean that
+            // the circuit went offline while sending the data, so simply wait until the
+            // client reconnects back or the circuit gets evicted because it stayed
+            // disconnected for too long.
+
+            var currentAttempt = 0;
+            do
             {
-                _client.SendAsync("JS.RenderBatch", _id, renderId, batchBytes).ContinueWith(sendTask =>
+                try
                 {
-                    if (sendTask.IsFaulted)
+                    if (!_client.Connected)
                     {
-                        pendingRenderInfo.TrySetException(sendTask.Exception);
+                        // If we detect that the client is offline. Simply stop trying to send the payload.
+                        // When the client reconnects we'll resend it.
+                        return;
                     }
-                });
-            }
-            catch (Exception syncException)
-            {
-                pendingRenderInfo.TrySetException(syncException);
-            }
 
-            // When the render is completed (success, fail, or timeout), stop tracking it
-            return pendingRenderInfo.Task.ContinueWith(task =>
-            {
-                _pendingRenders.TryRemove(renderId, out var ignored);
-                if (task.IsFaulted)
+                    if (!PendingRenderBatches.TryPeek(out var next) || pending.BatchId < next.BatchId)
+                    {
+                        // The client has ack this or a later batch already.
+                        return;
+                    }
+
+                    Log.BeginUpdateDisplayAsync(_logger, _client.ConnectionId);
+                    await _client.SendAsync("JS.RenderBatch", Id, pending.BatchId, pending.Data);
+                    await Task.Delay(SendMessageAcknowledgeInterval);
+                }
+                catch (Exception e)
                 {
-                    UnhandledException?.Invoke(this, task.Exception);
+                    Log.SendBatchDataFailed(_logger, e);
+                    if (currentAttempt > 0)
+                    {
+                        // Wait 10 seconds after we tried to send the payload, then check that that the client is still connected and
+                        // that we haven't received any ack from the client in the mean-time to retry.
+                        await Task.Delay(SendMessageRetryInterval);
```

Comment:
Could you clarify the sequence of events that leads up to us wanting to delay 10 seconds before retrying?

I would have guessed that we only need to send in two cases:

 * When a new batch is created and the client is connected, send that batch immediately
 * When the client (re)establishes connection, send all queued batches immediately

I guess there must be some other case I'm missing, but what is it? Waiting 10 seconds to see a UI update seems like it could be problematic for the user.

Reply:
This is done only on errors. When the send is successful we wait for a bit for the client to ack before we try to send the batch again (to avoid waisting network).

We always send the batches immediately there is only a delay when we retry, for example if we don't receive a confirmation from the client that the batch was rendered or the connection gets cut while sending the batch.

if we don't hear from the client we give up and we wait until it reconnects to send the batches (inmediately) or the circuit gets disposed.

Reply:
Note: we discussed and agreed to drop the retry logic in this PR, because once we place limits on the size of the queue of batches to send, it will no longer be useful.

Reply:
For the record, with the new behavior what will happens is the following:
If we fail to send a batch we will assume that the client disconnected. If the batch arrived to the client it will render it and send an ack, so the UI will be up to date and then two things can happen:
* A new batch is queued based on the state of the components (something updating data in real time)
* The client after reconnecting dispatches an event.

In both cases we would:
1) Resend all pending batches upon reconnect.
  * The client ignores batches it has already processed.
2) Send any newly produced batch right away.
  * The client will send an ack for the new batch and we will remove all pending batches from the queue.

The client did not receive the batch:
* The state of the UI would be the one for the previous batch (+ whatever disconnection logic is on the client).
* The client might take actions based on stalled UI (clicking a submit button twice if the connection goes away while sending the batch with the update). There would be a potential glitch in the UI, but not something that can't happen in any other UI framework.

As mentioned by @SteveSandersonMS. In the future we will place limits on the size of the queued batches and greatly limit the amount of work that we will do while the client is disconnected, so this won't be an issue.

Reply:
> The client might take actions based on stalled UI

As I understand it, that's no different from the inherent asynchrony in remote rendering. It's already possible to click a submit button twice (for example) if you do so faster than the whole async round trip. It's already up to the developer to put in whatever guards they want around actions that must be disabled.

Reply:
> As I understand it, that's no different from the inherent asynchrony in remote rendering.

Yep. That's what I'm trying to imply. That we are not creating a new problem for ourselves nor making an existing problem bigger.